Preprocess work

mikhailramalho opened the pull request:
- Update psyco set file to point to the right files
- New script to preprocess files
- Update preprocessed files from the BitVectorsReach category
- Update preprocessed files from the Floats category

PhilippWendler commented:
Why do `.set` files need to be changed when fixing pre-processed benchmarks?

Please split this into smaller pieces. I would suggest to have one pull request per directory.

This needs very careful reviewing to make sure we do not override any changes to the pre-processed files or inadvertently change the semantics of some files.
Thus I would prefer to have the changes as easily reviewable as possible. For example when I did these fixes for HeapManipulation in #144, I split the changes into several commits. Because all the whitespace-related and trivial changes were in separate commits, it was much easier to check the actual changes from 5925893bbf33fe00dddc0ad9334404d24be37b38 so see whether they change something relevant.

mikhailramalho commented:
>
> Why do .set files need to be changed when fixing pre-processed benchmarks?
>
For most of them, the set files were pointing to the *.c files, I changed
them to point to the *.i files.

Specifically for ProductLines.set, commit 29d7ee946 changed the file names
and the set files was not finding the benchmarks anymore.

Please split this into smaller pieces. I would suggest to have one pull
> request per directory.
>
> This needs very careful reviewing to make sure we do not override any
> changes to the pre-processed files or inadvertently change the semantics of
> some files.
> Thus I would prefer to have the changes as easily reviewable as possible.
> For example when I did these fixes for HeapManipulation in #144
> <https://github.com/sosy-lab/sv-benchmarks/pull/144>, I split the changes
> into several commits. Because all the whitespace-related and trivial
> changes were in separate commits, it was much easier to check the actual
> changes from 5925893
> <https://github.com/sosy-lab/sv-benchmarks/commit/5925893bbf33fe00dddc0ad9334404d24be37b38>
> so see whether they change something relevant.
>

I've done it at first, but trust me, it doesn't make it any easier. My
suggestion is to use github for revision, the side-by-side comparison is
great to check the diffs (you'll want to skip the headers expansion anyway
as they are the biggest part of the changes and aren't really that
important).

Thank you,

-- 

Mikhail Ramalho.

PhilippWendler commented:
I would not add `.i` files for `.c` files where this is not necessary. Having the same file twice is just redundant and potentially confusing.

The problem with `ProductLines.set` is already fixed.

I can't even look at all the changes in GitHub because it just tells me that there are too many. Furthermore, there are a lot of trivial and whitespace changes, so I would fear that any relevant changes get lost in these.

Anyway, there seem to be manual changes mixed into this pull request. Even if we agree on having the preprocessing changes as a single large chunk, any manual changes really must be made separately.

mikhailramalho commented:
Alright, I updated the pull request:

* Added the preprocess script
* Fix benchmarks that were preprocessed with pragma lines in Floats category and HeapReach category.

Fun fact: most of the benchmarks in the repo were preprocessed using gcc, while the ones in c/forester-heap/ were preprocessed using clang.

PhilippWendler commented:
Sorry, when reading your [comment](https://github.com/sosy-lab/sv-benchmarks/pull/169#issuecomment-257883541) I did not realize that you had completely replaced the contents of this pull request and thought it was still based on the large set of changes that it initially contained (which I did not want to review). Maybe it is better in such a case to open a new pull request to avoid confusion? In fact, while I am typing this comment because you asked for reviews on the list GitHub showed me that the content of the pull request has changed again and I am wondering whether this is now the final state that should be reviewed or not.

Anyway, for the duplicate work on HeapReach I am sorry, this was because I did not notice that your PR contained this as explained above. However, I prefer the less intrusive changes of just removing the line directives than a fully new pre-processing anyway.

Now the current state of the PR contains a commit by leostrakosch? Why this? I keep getting more and more confused. The content of the PR also does not match its description at all anymore. I suggest to close this PR, and open separate PRs for each block of changes that you want to make.

For the script that you propose, what is actually its use case?

For submitters of new benchmarks this script is not good because it would re-preprocess all other files of the same category, and we certainly do not want such changes mixed with new benchmark submissions. So having this script in the repository with a name that suggests that people should use it could easily make things worse.

For fixing existing benchmarks I don't think we would need to commit a script, because this is a one-time effort. And basically the script just replaces a shell one-liner. And the default in the script is wrong: if almost everything in the repository was preprocessed with GCC, we should certainly keep it as it is (assuming this makes any difference at all, otherwise why even have the choice?).

Also the script still relies on people manually specifying the correct architecture, so it is not less error-prone than manual preprocessing. The script would actually have some real benefit if it read the correct architecture from the new category configuration files (but the above questions about its use cases would still remain).

mikhailramalho commented:
Hi Phillip,

No problem, I just found it odd the duplicated work.

The idea is to use the script to check for inconsistency. I added two cases
where it found wrongly preprocessed categories, Floats and HeapReach, but
as it stands now, it's only the script.

The idea is to ask developers to run the script, after committing the
benchmarks (before the pull request). It will also make it easier for
anyone to review if they were correctly preprocessed, so it actually makes
the whole process less error-prone before accepting the merge: both the
submitter and reviewer(s) would have to set the wrong flags to get the
wrong result.

I agree that this pull request is a bit messy though. I'll close it and
create another. After that, I can go back to check other categories.



2016-11-05 12:56 GMT+00:00 Philipp Wendler <notifications@github.com>:

> Sorry, when reading your comment
> <https://github.com/sosy-lab/sv-benchmarks/pull/169#issuecomment-257883541>
> I did not realize that you had completely replaced the contents of this
> pull request and thought it was still based on the large set of changes
> that it initially contained (which I did not want to review). Maybe it is
> better in such a case to open a new pull request to avoid confusion? In
> fact, while I am typing this comment because you asked for reviews on the
> list GitHub showed me that the content of the pull request has changed
> again and I am wondering whether this is now the final state that should be
> reviewed or not.
>
> Anyway, for the duplicate work on HeapReach I am sorry, this was because I
> did not notice that your PR contained this as explained above. However, I
> prefer the less intrusive changes of just removing the line directives than
> a fully new pre-processing anyway.
>
> Now the current state of the PR contains a commit by leostrakosch? Why
> this? I keep getting more and more confused. The content of the PR also
> does not match its description at all anymore. I suggest to close this PR,
> and open separate PRs for each block of changes that you want to make.
>
> For the script that you propose, what is actually its use case?
>
> For submitters of new benchmarks this script is not good because it would
> re-preprocess all other files of the same category, and we certainly do not
> want such changes mixed with new benchmark submissions. So having this
> script in the repository with a name that suggests that people should use
> it could easily make things worse.
>
> For fixing existing benchmarks I don't think we would need to commit a
> script, because this is a one-time effort. And basically the script just
> replaces a shell one-liner. And the default in the script is wrong: if
> almost everything in the repository was preprocessed with GCC, we should
> certainly keep it as it is (assuming this makes any difference at all,
> otherwise why even have the choice?).
>
> Also the script still relies on people manually specifying the correct
> architecture, so it is not less error-prone than manual preprocessing. The
> script would actually have some real benefit if it read the correct
> architecture from the new category configuration files (but the above
> questions about its use cases would still remain).
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/sosy-lab/sv-benchmarks/pull/169#issuecomment-258609850>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABREnyoSbbpdHkk_LaxKfuoWhm89Lda4ks5q7Hz9gaJpZM4KjeaI>
> .
>



-- 

Mikhail Ramalho.

PhilippWendler commented:
A script that can detected inconsistencies in pre-processing would be great! But I must be missing something, how can your script do this? I just find that I can use it for pre-processing like I could do manually.

For pull requests, I think running this script is not a good idea: Because it would pre-process even already existing files, we would get a lot of unwanted changes mixed into the new benchmarks.
